modules/core/src/main/scala/fs2/kafka/KafkaProducer.scala

```diff
@@ -117,12 +119,11 @@ object KafkaProducer {
                 callback { (metadata, exception) =>
                   val complete =
                     deferred.complete {
-                      if (exception == null)
-                        Right((record, metadata))
+                      if (exception == null) Right((record, metadata))
                       else Left(exception)
                     }
 
-                  F.runAsync(complete)(_ => IO.unit).unsafeRunSync()
+                  dispatcher.unsafeRunSync(complete.void)
```

Comment:
I'm not sure what the reason for using `runAsync` was here, so not sure whether this will work

Reply:
That's because `F.async` will wait for the callback call and will not release `F` from the whole `commitAsync` method. But in the java client, a callback for the `commitAsync` could be resolved only with the `poll` calls. So, to make it work we need to release `commitAsync` right after the call and call `poll` after that. This is not how `F.async` works.

Reply:
I'll need to take a while to get my head around that, but thanks 🙂

Reply:
I think Ben is asking about `runAsync` vs `unsafeRunSync` here, rather than `async` vs `delay`, but you make a good point regardless.
Is there a reason why this cannot be rewritten as
```
F.async_ { cb =>
  producer.send(
        javaRecord,
        callback { (metadata, exception) =>
              if (exception == null) cb(Right((record, metadata)))
              else cb(Left(exception))
         })
     }.start.map(_.join)
```

Reply:
I tried and it just didn't work for me. I didn't make a deep investigation though, so maybe I was doing something wrong. Theoretically, it should work. But callback just doesn't invoked for me when I'm trying it. Maybe because `poll` and this callback should be done in the same thread (that's how java client works)?

Reply:
@LMnet when you tried this did you do `.start.map(_.join)`? Because I previously tried changing the CE2 code to use `async` but without `start` and the tests froze like you're reporting, but now I'm trying it with `start` and it seems to work.

Reply:
If that does work, hopefully we can remove the `ConcurrentEffect` requirement with CE2.

Reply:
`joinAndEmbedNever` is the CE3 equivalent

Reply:
Unfortunately, I don't remember exactly. I think I tried `.start.map(_.join)`, but I may be wrong.

Nevertheless, if it works — it's ok. However, to be completely sure, I would check that the thread which calls `callback` is the same thread, that calls `poll`.

Reply:
>  I would check that the thread which calls callback is the same thread, that calls poll.

do you have any links to code or doc that specify this? I haven't seen it as a requirement for this method (although other parts of Kafka do require it)

Reply:
> I would check that the thread which calls `callback` is the same thread, that calls `poll`.

Which `poll` are we talking about? I can only find references to `poll` in the consumer, not the producer.

Reply:
> I can only find references to poll in the consumer, not the producer.

that's my impression as well

Reply:
Oh, guys, I'm sorry, but I thought we are talking about the `KafkaConsumer` It has almost the same pattern with callback in the `commitAsync` method.

Reply:
ah cool, worth keeping that comment in mind for the consumer anyway :)

Reply:
Does that mean we could make a similar change in the consumer if we run `poll` and the callback on a dedicated single-thread execution context?

Reply:
Specifically, I was talking about this: https://github.com/fd4s/fs2-kafka/pull/417/files?file-filters%5B%5D=.scala#diff-bcbdc586bf79fc42fcb6c0c4e87ec601c395715522f3af680cf988c36cc4d1eaR212

Reply:
> Does that mean we could make a similar change in the consumer if we run poll and the callback on a dedicated single-thread execution context?
It would be really cool, because atm there are a few `unsafeRunSync` calls in the consumer. I will be happy to get rid of them if possible.

Reply:
I've opened a PR that changes this on the producer side - https://github.com/fd4s/fs2-kafka/pull/418/files#diff-c4655864f9261fc7a30abdbdd8df44a5d4ed192dcdba30f5ab511b4f7dc8cb8f

Reply:
> @LMnet I think Ben is asking about `runAsync` vs `unsafeRunSync` here, rather than `async` vs `delay`, but you make a good point regardless.
> Is there a reason why this cannot be rewritten as
> 
> ```
> F.async_ { cb =>
>   producer.send(
>       javaRecord,
>       callback { (metadata, exception) =>
>          if (exception == null) cb(Right((record, metadata)))
>          else cb(Left(exception))
>         }
>      )
>   }.start.map(_.join)
> ```

As it turns out, yes: https://github.com/fd4s/fs2-kafka/issues/477